File: platform/extension-sdk/python/src/aks_platform_sdk/key_vault.py
from dataclasses import dataclass
from typing import Any

from .credential import create_default_credential


@dataclass
class KeyVaultClient:
    vault_url: str
    credential: Any | None = None

    def __post_init__(self) -> None:
        if not self.vault_url.strip():
            raise ValueError("Key Vault URL is required.")

        from azure.keyvault.secrets import SecretClient

        self._client = SecretClient(
            vault_url=self.vault_url,
            credential=self.credential or create_default_credential(),
        )

    def get_secret(self, name: str) -> str | None:
        return self._client.get_secret(name).value

    def get_required_secret(self, name: str) -> str:
        value = self.get_secret(name)
        if not value:
            raise ValueError(f"Secret '{name}' was found but has no value.")
        return value

File: platform/extension-sdk/node/src/key-vault.ts
import { createDefaultCredential, type TokenCredentialLike } from "./credential.js";

export interface KeyVaultClientOptions {
  vaultUrl: string;
  credential?: TokenCredentialLike;
}

export interface KeyVaultClient {
  getSecret(name: string): Promise<string | undefined>;
  getRequiredSecret(name: string): Promise<string>;
}

export async function createKeyVaultClient(options: KeyVaultClientOptions): Promise<KeyVaultClient> {
  if (!options.vaultUrl || options.vaultUrl.trim().length === 0) {
    throw new Error("Key Vault URL is required.");
  }

  const credential = options.credential ?? await createDefaultCredential();
  const keyVault = await import("@azure/keyvault-secrets");
  const client = new keyVault.SecretClient(options.vaultUrl, credential);

  return {
    async getSecret(name: string): Promise<string | undefined> {
      const secret = await client.getSecret(name);
      return secret.value;
    },

    async getRequiredSecret(name: string): Promise<string> {
      const secret = await client.getSecret(name);
      if (!secret.value) {
        throw new Error(`Secret '${name}' was found but has no value.`);
      }
      return secret.value;
    }
  };
}
